README
Mail-DKIM version 0.41
======================

This module implements the various components of the DKIM and DomainKeys
message-signing and verifying standards for Internet mail. It currently
tries to implement these specifications:
 * RFC4871, for DKIM
 * RFC4870, for DomainKeys

With each release, this module is getting bigger, but don't worry,
most of the growth is from having more things to test with `make test'.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Crypt::OpenSSL::RSA
  Digest::SHA
  Mail::Address (part of the MailTools package)
  MIME::Base64
  Net::DNS

USAGE

Decide whether you want to "sign" or "verify" messages.
To sign, see the Mail::DKIM::Signer module.
To verify, see the Mail::DKIM::Verifier module.

BUGS

Some details of the specification are not completely implemented.
See the TODO file for a list of things I know about.

Report bugs to Jason Long <jason@long.name>.
If `make test' fails, please include the versions of your
installed Crypt::OpenSSL::RSA module and OpenSSL libraries.

COPYRIGHT AND LICENCE

Copyright (C) 2010 by Jason Long
Copyright (C) 2006-2009 by Messiah College

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

SEE ALSO

The DKIM proxy home page, http://dkimproxy.sourceforge.net/.


Changes
This file summarizes what's changed between releases of Mail-DKIM.
See the ChangeLog file for the details.

Version 0.41

  * Fix to better comply with DKIM specification:
    * DKIM signatures with multiple methods specified in q= tag are now
      accepted. Such signatures are not known to appear in real life.
      (rt.cpan.org issue #83176)


Version 0.40 - released 2013-02-07

  * New/changed functionality:
    * a single DNS resolver is created for the lifetime of the program,
      rather than reinitializing the resolver for each new query.

  * bugfixes:
    * fix the error message given when an invalid algorithm is
      specified in the construction of Mail::DKIM::Signer.
    * avoid Perl warning about use of an undefined value in several
      places (rt.cpan.org issue #82913).
    * speed- improved performance of parsing the message into lines
      (rt.cpan.org issue #77902). Patch by Mark Martinec.
    * fix DNS queries to use the correct method (txtdata) of Net::DNS
      (rt.cpan.org issue #83170). Patch by Mark Martinec.

  * API changes:
    * global subroutines resolver() or enable_EDNS0() in module
      Mail::DKIM::DNS can be called to specify non-default options
      to Net::DNS::Resolver (see also rt.cpan.org issue #80425).


Version 0.39 - released 2010-11-14

  * bugfixes:
    * fix issue with getting wrong error codes when q= tag is empty
      (issue #3011005)
    * anti-abuse- prevent a message with thousands of signatures from
      thrashing the whole computer (issue #3010997)
    * memory usage- significantly reduced memory footprint for
      processing a message with a large header and many signatures
    * fix error message given when no KeyFile has been specified
      (issue #1889690)

  * API changes:
    * the Canonicalization::finish_header() method now expects a
      argument to be passed to it. In the unusual case that you are
      using this method from your own code, please update your code.


Version 0.38 - released 2010-03-31

  * New/changed functionality:
    * DNS lookups can now be started asynchronously; the queries are
      created as the header is parsed; the results are not actually
      needed until the entire message has been read. (The Mail::DKIM
      module does not yet do the queries asynchrously; this is just
      the infrastructure so that the queries can be asynchronous in
      the future.)

  * bugfixes:
    * DNS lookup overrides alarm() signal (issue #2854325)

  * documentation updates:
    * document use of custom PrivateKey object, for external signing
    * describe how to get "pretty signatures" in Signer.pm


Version 0.37 - released 2009-09-08

  * New/changed functionality:
    * ADSP records now check whether the domain itself exists, in
      accordance to the ADSP specification

  * bugfixes:
    * fixed regexp used to detect header field names (issue #2803465)
    * various fixes to ADSP checking


Version 0.36 - released 2009-06-02

  * API changes:
    * restore the as_string() method which was accidentally removed
      in version 0.34


Version 0.35 - released 2009-05-22

  * bugfixes:
    * fixed a runaway regular expression in the canonicalization
      routines (patch provided by Mark Martinec)


Version 0.34 - released 2009-05-20

  * New/changed functionality:
    * support for ADSP (author-domain-signing-practices) records
    * removed support for pre-standardized DKIM signatures (i.e. these
      are DKIM signatures without a v= or bh= tag).
    * DNS resolver errors are detected and reported as such

  * API changes:
    * renamed Mail::DKIM::Policy to Mail::DKIM::DkPolicy. Programs using
      the former name to create policy objects directly (though it would
      be more expected to fetch the objects through
      Mail::DKIM::Verifier) should update their code
    * new policies() method in Mail::DKIM::Verifier for fetching all
      applicable sender/author signing policies

  * bugfixes:
    * Signer object would die if first line of input wasn't a header
      (rt.cpan.org issue #46179)


Version 0.33 - released 2009-03-10

  * bugfixes:
    * signature wrapping would sometimes cause improper preparation of
      DKIM signatures, with "simple" canonicalization (issue #2257046)
  * test scripts:
    * the included corpus is now verified using a fake-DNS resolver,
      which means the test corpus can validate even when your DNS
      servers are really slow


Version 0.32 - released 2008-06-03

  * removed requirement for Digest::SHA1 (issue #1832549).
    We now use the more capable Digest::SHA module for SHA-1 and SHA-256.
  * bugfixes:
    * granularity checking should be case-sensitive (issue #1938112).
    * identity tag now uses quoted-printable encoding (issue #1839015).
  * API improvement:
    * implemented identity_source() for DkSignature objects 


Version 0.31 - released 2008-04-14

  * some error detail messages were changed (see ChangeLog, 2008-01-10 entry)
  * by default, and when possible, DKIM signatures now omit c= and q= tags
    (they are optional tags) (issue #1878518)
  * DKIM and DomainKey signatures are now wrapped so that line breaks
    occur before colon (':') separators instead of after; this avoids
	confusing some broken MUAs (issue #1868648)
  * bugfixes:
    * "undef value" error when DKIM signature appears at end of header
      (issue #1878954)
    * use proper regexp for splitting email address (issue #1878994)
  * API improvements:
    * can specify a domain for fetch_author_policy() (issue #1879197)
    * can access a signature's public-key object (issue #1879215)
    * can specify an OpenSSL-private-key object for PrivateKey->new()
      (issue #1879209)


Version 0.30.1 - released 2008-01-24

 * bugfix:
   * email from cisco.com was failing to verify (issue #1878523)


Version 0.30 - released 2008-01-10

 * includes speed-up optimizations by Mark Martinec
 * DomainKeys, implement proper identity matching...

lib/Mail/DKIM/Key.pm
#!/usr/bin/perl
#
# Copyright 2006 Jason Long. All rights reserved.
#
# Copyright (c) 2004 Anthony D. Urso. All rights reserved.
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.

package Mail::DKIM::Key;

use strict;

sub cork {
	my $self = shift;

	(@_) and
		$self->{'CORK'} = shift;

	$self->{'CORK'} or
		$self->convert;

	$self->{'CORK'};
}

sub data {
	my $self = shift;

	(@_) and 
		$self->{'DATA'} = shift;

	$self->{'DATA'};
}

sub errorstr {
	my $self = shift;

	(@_) and 
		$self->{'ESTR'} = shift;

	$self->{'ESTR'};
}

sub size {
	my $self = shift;

	return $self->cork->size * 8;
}

sub type {
	my $self = shift;

	(@_) and 
		$self->{'TYPE'} = shift;

	$self->{'TYPE'};
}

use Crypt::RSA::DataFormat ("h2osp");

sub calculate_EM
{ 
	my ($digest_algorithm, $digest, $emLen) = @_;

	# this function performs DER encoding of the algorithm ID for the
	# hash function and the hash value itself
	# It has this syntax:
	#      DigestInfo ::= SEQUENCE {
	#          digestAlgorithm AlgorithmIdentifier,
	#          digest OCTET STRING
	#      }

	# RFC 3447, page 42, provides the following octet values:
	my %digest_encoding = (
		"SHA-1" => h2osp(
			"0x 30 21 30 09 06 05 2B 0E 03 02 1A 05 00 04 14"),
		"SHA-256" => h2osp(
			"0x 30 31 30 0d 06 09 60 86 48 01 65 03 04 02 01 05 00 04 20"),
		);

	defined $digest_encoding{$digest_algorithm}
		or die "Unsupported digest algorithm '$digest_algorithm'";

	my $T = $digest_encoding{$digest_algorithm} . $digest;
	my $tLen = length($T);

	if ($emLen < $tLen + 11)
	{
		die "Intended encoded message length too short.";
	}

	my $PS = chr(0xff) x ($emLen - $tLen - 3);
	my $EM = chr(0) . chr(1) . $PS . chr(0) . $T; 
	return $EM;
}

1;
